text/0000-futures.md

```diff
@@ -154,31 +164,316 @@ can be dropped.
+pub struct LocalWaker {
+    waker: RawWaker,
+}
+
+impl LocalWaker {
+    /// Wake up the task associated with this `LocalWaker`.
+    pub fn wake(&self) {
+        // The actual wakeup call is delegated through a virtual function call
+        // to the implementation which is defined by the executor
+        unsafe { (self.waker.vtable.wake)(self.waker.data) }
+    }
+
+    /// Converts the `LocalWaker` into `Waker`, which can be sent across
+    /// thread boundaries.
+    ///
+    /// This operation consumes the `LocalWaker`.
+    ///
+    /// This function can panic if the associated executor does not support
+    /// getting woken up from a different thread.
+    pub fn into_waker(self) -> Waker {
+        match self.try_into_waker() {
+            Some(waker) => waker,
+            None => panic!("Conversion from LocalWaker into Waker is not supported"),
+        }
+    }
+
+    /// Tries to convert the `LocalWaker` into a `Waker`, which can be sent
+    /// across thread boundaries.
+    ///
+    /// This operation consumes the `LocalWaker`.
+    ///
+    /// Returns None if the if the associated executor does not support
+    /// getting woken up from a different thread.
+    pub fn try_into_waker(self) -> Option<Waker> {
+        unsafe {
+            let maybe_raw_waker = (self.waker.vtable.into_waker)(self.waker.data);
+            // Avoid that the drop runs on self, which would e.g. decrease
+            // the refcount on it. The object has been consumed or converted
+            // by the `into_waker` function.
+            mem::forget(self);
+            match maybe_raw_waker {
+                Some(rw) => Some(Waker::new_unchecked(rw)),
+                None => None,
+            }
+        }
+    }
+
+    /// Returns whether or not this LocalWaker and other LocalWaker awaken the same task.
+    ///
+    /// This function works on a best-effort basis, and may return false even
+    /// when the LocalWakers would awaken the same task. However, if this function
+    /// returns true, it is guaranteed that the LocalWakers will awaken the same task.
+    ///
+    /// This function is primarily used for optimization purposes.
+    pub fn will_wake(&self, other: &LocalWaker) -> bool {
+        self.waker == other.waker
+    }
+
+    /// Creates a new `LocalWaker` from `RawWaker`.
+    ///
+    /// The method can not check whether `RawWaker` fulfills the required API
+    /// contract to make it usable for `LocalWaker` and is therefore unsafe.
+    pub unsafe fn new_unchecked(waker: RawWaker) -> LocalWaker {
+        LocalWaker {
+            waker: waker,
+        }
+    }
+}
+
+// The implementations of `Clone` and `Drop` follow what is shown in for `Waker`.
+```
+
+`Waker`s must fulfill the following requirements:
+- They must be cloneable
+- If all instances of a `Waker` have been dropped and their associated task had
+  been driven to completion, all resources which had been allocated for the task
+  must have been released.
```

Comment:
Why is this a requirement? I don't disagree with the general logic behind it, I'm just wondering why this is a requirement.

Reply:
While I agree that in practice basically every `Waker` should clean up after itself, I do think calling it a requirement gives the wrong impression. This implies (to me) that a lack of resource leaks is a *safety* property that code can rely on for memory safety. Along the same lines that `std::mem::forget` is safe but not recommended, I think that cleaning up can be mentioned, but should not be a requirement.